### PR TITLE
[freertos] Fix freertos linker error

### LINKED ISSUE #523: Linker error FreeRTOSConfig.h with gcc 10
Linking examples fails because of [`modm_extern_c uint32_t _ZN4modm8platform4fcpuE;`](https://github.com/modm-io/modm/blob/develop/ext/aws/FreeRTOSConfig.h.in#L39) in FreeRTOSConfig.h
When compiling the header as C, `modm_extern_c` is defined as empty. Thus, we have a **definition** inside a header file. If we include this file in multiple C translation units, there will be multiple definitions and linking fails. That is the reason we see linker errors for freertos C files but not for C++ sources.

```
ld: build/release/modm/libmodm.a(list.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
ld: build/release/modm/libmodm.a(port.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
ld: build/release/modm/libmodm.a(queue.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
ld: build/release/modm/libmodm.a(stream_buffer.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
ld: build/release/modm/libmodm.a(tasks.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
ld: build/release/modm/libmodm.a(timers.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/stm32f746g_discovery/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
```
I can't actually find a definition for it outside of the header in the generated code for STM32.

Apart from that, declaring a C++ variable as extern "C" with its mangled name is at least confusing and implementation defined. Maybe we can find a cleaner way to achieve the same result.

### PR DESCRIPTION
**Pull request by chris-durand**
Closes #523

### CODE REVIEW

**Review by salkinium — APPROVED** [review]
Thanks! I have no clue why this worked before…

**Comment by rleh** [comment]
... but the errors are still there?!

**Comment by salkinium** [comment]
crap

**Comment by salkinium** [comment]
It's been so long since I merged stuff, I forgot to check the CI. Dayum…

File: ext/aws/FreeRTOSConfig.h.in [hunk 1]
```diff
@@ -36,12 +36,12 @@ your application. */
 #include <modm/platform/device.hpp>
 
 // declared in modm:platform:clock
-modm_extern_c uint32_t _ZN4modm8platform4fcpuE;
+modm_extern_c uint32_t SystemCoreClock;
```
Comment (salkinium) [inline ext/aws/FreeRTOSConfig.h.in:39]: Should `modm_extern_c` really be empty for C? I feel like it should be `extern "C"` for C++ and `extern` for C.

Comment (rleh) [reply ext/aws/FreeRTOSConfig.h.in:39]: Maybe we should completely remove the `modm_extern_c` macro to prevent such confusion in future?

Comment (salkinium) [reply ext/aws/FreeRTOSConfig.h.in:39]: Yeah, I'm not impressed by it either…

Comment (salkinium) [reply ext/aws/FreeRTOSConfig.h.in:39]: It's used in conflicting ways as a function attribute and a variable attribute:

```
modm/board/board.cpp.in
29:modm_extern_c void

modm/architecture/interface/assert.hpp.in
172:modm_extern_c void

modm/architecture/interface/interrupt.hpp
98:		modm_extern_c void vector ## _vect(void)
109:		modm_extern_c void vector ## _IRQHandler(void)
113:		modm_extern_c void vector ## _IRQHandler(void) \
122:		modm_extern_c void vector ## _isr(void)
126:		modm_extern_c void vector ## _isr(void) __VA_ARGS__; \

modm/architecture/interface/assert.h.in
109:modm_extern_c void

modm/architecture/utils.hpp
152:	#	define modm_extern_c extern "C"
154:	#	define modm_extern_c
170:		modm_extern_c uint8_t name ## _laddr[];		\
171:		modm_extern_c const size_t name ## _length = (size_t) name ## _laddr; \
172:		modm_extern_c uint8_t name[]
```

Comment (salkinium) [reply ext/aws/FreeRTOSConfig.h.in:39]: | type | C | C++ |
|---|:---|:---|
| function | `void fn()` | `extern "C" void fn()` |
| variable | `extern int var;` | `extern "C" int var` |

Help?

Comment (salkinium) [reply ext/aws/FreeRTOSConfig.h.in:39]: So I have to have `modm_extern_c_function` and `modm_extern_c_variable` macros? WTF is this…

Comment (chris-durand) [reply ext/aws/FreeRTOSConfig.h.in:39]: Oh, yes this was only the first part of the fix. The `extern` problem remains.

Comment (chris-durand) [reply ext/aws/FreeRTOSConfig.h.in:39]: @salkinium You could try to unify the macros. Putting `extern` on a function declaration does nothing in most cases, external linkage is the default.

**Pull request merged**